src/main/java/org/jenkinsci/remoting/engine/JnlpServer3Handshake.java

```diff
@@ -95,11 +95,25 @@ public Channel connect() throws IOException, InterruptedException {
                 Jnlp3Util.keyFromString(aesKeyString),
                 Jnlp3Util.keyFromString(specKeyString));
 
-        Channel channel = createChannelBuilder(nodeName).build(
-                new CipherInputStream(SocketChannelStream.in(socket),
-                        channelCiphers.getDecryptCipher()),
-                new CipherOutputStream(SocketChannelStream.out(socket),
-                        channelCiphers.getEncryptCipher()));
+
+        CipherInputStream in = null;
+        CipherOutputStream out = null;
+        final Channel channel;
+        try {
+            in = new CipherInputStream(SocketChannelStream.in(socket), channelCiphers.getDecryptCipher());
+            out = new CipherOutputStream(SocketChannelStream.out(socket), channelCiphers.getEncryptCipher());
+            channel = createChannelBuilder(nodeName).build(in, out);
+        } catch (Exception ex) {
+            // Something went wrong. We want to gracefully close Cipher stream in the case they were open
+            EngineUtil.closeAndLogFailures(in, LOGGER, "CipherInputStream", socket.toString());
+            EngineUtil.closeAndLogFailures(out, LOGGER, "CipherOutputStream", socket.toString());
```

Comment:
Even better would be a utility (should I file a suggestion for Guava?):

```java
public static class ClosePartials implements AutoCloseable {
    private final List<AutoCloseable> things = new ArrayList<>();
    public ClosePartials() {}
    public <T extends AutoCloseable> T with(T thing) {
        things.add(thing);
        return thing;
    }
    /** @param <T> recommended to also be {@link AutoCloseable} */
    public <T> T done(T thing) {
        things.clear();
        return thing;
    }
    @Override
    public void close() throws Exception {
        Exception x = null;
        for (AutoCloseable thing : things) {
            try {
                thing.close();
            } catch (Throwable t) {
                if (x == null) {
                    if (t instanceof Exception) {
                        x = (Exception) t;
                    } else {
                        x = new Exception(t);
                    }
                } else {
                    x.addSuppressed(t);
                }
            }
        }
        if (x != null) {
            throw x;
        }
    }
}
```

Example usage:

```java
public static Channel example(Socket s, Cipher c, ChannelBuilder cb) throws Exception {
    try (ClosePartials cp = new ClosePartials();
         InputStream sin = cp.with(SocketChannelStream.in(s));
         CipherInputStream cin = cp.with(new CipherInputStream(sin, c));
         OutputStream sout = cp.with(SocketChannelStream.out(s));
         CipherOutputStream cout = cp.with(new CipherOutputStream(sout, c))) {
        return cp.done(cb.build(cin, cout));
    }
}
```

In general, methods which deliberately create closeable _[sic]_ objects but then conceal them as references inside some higher-level closeable object are risky, because even if the ultimate caller is careful to `close` the final object, its `close` method might not be foolproof. `Channel.close` does not even pretend to guarantee that all streams it was originally passed will be closed by the time the method completes.